metis-media-service/src/main/java/eu/europeana/metis/mediaservice/MediaProcessor.java

```diff
@@ -0,0 +1,441 @@
+	private static final String[] THUMB_SUFFIX = { "-MEDIUM", "-LARGE" };
+	
+	private static File tempDir;
+	private static File colormapFile;
+	private static String magickCmd;
+	private static String ffprobeCmd;
+	private static Tika tika;
+	
+	private ExecutorService commandIOThreadPool = Executors.newFixedThreadPool(2);
+	
+	private EdmObject edm;
+	private Map<String, List<UrlType>> urlTypes;
+	private Map<File, String> thumbnails = new HashMap<>();
+	
+	public MediaProcessor() {
+		staticInit();
+	}
+	
+	/**
+	 * @param edm future calls to {@link #processResource(String, String, File)}
+	 *        will store extracted metadata in given EDM.
+	 */
+	public void setEdm(EdmObject edm) {
+		this.edm = edm;
+		urlTypes = edm.getResourceUrls(Arrays.asList(UrlType.values()));
+		thumbnails.clear();
+	}
+	
+	public EdmObject getEdm() {
+		return edm;
+	}
+	
+	/**
+	 * @param contents downloaded file, can be {@code null} for mime types accepted
+	 *        by {@link #supportsLinkProcessing(String)}
+	 */
+	public void processResource(String url, String providedMimeType, File contents) throws MediaException {
+		String mimeType;
+		try {
+			mimeType = contents != null ? tika.detect(contents) : tika.detect(URI.create(url).toURL());
+		} catch (IOException e) {
+			throw new MediaException("Mime type checking error", "IOException " + e.getMessage(), e, contents == null);
+		}
+		
+		if (!mimeType.equals(providedMimeType))
+			log.info("Invalid mime type provided (should be {}, was {}): {}", mimeType, providedMimeType, url);
+		if (contents == null && !supportsLinkProcessing(mimeType))
+			throw new IllegalArgumentException("Contents file is required for mime type " + mimeType);
+		
+		try {
+			if (isImage(mimeType))
+				processImage(url, mimeType, contents);
+			if (isAudioVideo(mimeType))
+				processAudioVideo(url, mimeType, contents);
+			if (isText(mimeType))
+				processText(url, mimeType, contents);
+		} catch (IOException e) {
+			throw new MediaException("I/O error during procesing", "IOException " + e.getMessage(), e);
+		}
+	}
+	
+	/**
+	 * @return thumbnails for all the image resources processed since the last call
+	 *         to {@link #setEdm(EdmObject)}. The map's key is the thumbnail file,
+	 *         with the name it should be stored under. The value is the resource's
+	 *         original url. Remember to remove the files when they are no longer
+	 *         needed.
+	 */
+	public Map<File, String> getThumbnails() {
+		return new HashMap<>(thumbnails);
+	}
+	
+	@Override
+	public void close() {
+		commandIOThreadPool.shutdown();
+	}
+	
+	private void processImage(String url, String mimeType, File content) throws MediaException, IOException {
+		List<File> thumbs = prepareThumbnailFiles(url, mimeType);
+		
+		int sizes = THUMB_SIZE.length;
+		ArrayList<String> command = new ArrayList<>(Arrays.asList(
+				magickCmd, content.getPath() + "[0]", "-format", "%w\n%h\n%[colorspace]\n", "-write", "info:"));
```

Comment:
Create constants with all the literals, especially the ones that indicate parameter names and/or are used more than once. Like "-write" or "application/pdf" for example.

Reply:
If "-write" happened to be one character shorter or used one time less, this rule would not trigger.
It wouldn't make sense to make this literal stand out by turning it into a constant. And adding a dozen constants would look even worse.